File: frontend/src/api/analysis.ts
import apiClient, { unwrap } from './client'

export interface KnowledgePoint {
  id: number
  name: string
  description?: string
  importance: string
  frequency: number
  detail?: KnowledgePointDetail
}

export interface KnowledgePointDetail {
  overview: string
  chapter_name?: string
  importance_label?: string
  linked_question_count?: number
  exam_focus: string[]
  answer_template: string[]
  common_mistakes: string[]
  study_advice: string[]
}

export interface KnowledgeTreeChapter {
  id: number
  name: string
  order: number
  description?: string
  points: KnowledgePoint[]
}

export interface KnowledgeTree {
  subject_id: number
  chapters: KnowledgeTreeChapter[]
}

export interface HighFrequencyPoint extends KnowledgePoint {
  trend: 'up' | 'stable' | string
  trend_slope?: number
  next_year_prediction?: number
  chapter_name?: string
  question_count?: number
  confidence?: number
  reason?: string
}

export interface TrendPoint {
  year: number
  count: number
}

export interface PointTrend {
  subject_id: number
  point_id: number
  values: TrendPoint[]
  trend?: 'up' | 'down' | 'stable' | string
  trend_slope?: number
  next_year_prediction?: number
}

export interface WordCloudItem {
  word: string
  weight: number
}

export interface KnowledgeNetworkNode {
  id: number
  name: string
  chapter_name?: string
  frequency: number
  trend?: string
  importance?: string
  question_count?: number
}

export interface KnowledgeNetworkEdge {
  source: number
  target: number
  weight: number
  source_name?: string
  target_name?: string
}

export interface KnowledgeNetwork {
  subject_id: number
  nodes: KnowledgeNetworkNode[]
  edges: KnowledgeNetworkEdge[]
}

export interface HotspotAlert {
  point_id: number
  name: string
  chapter_name?: string
  severity: 'high' | 'medium' | string
  trend: string
  trend_slope: number
  confidence: number
  priority_score: number
  message: string
}

export interface ChapterHeatmapItem {
  id: number
  name: string
  frequency: number
}

export interface ChapterHeatmap {
  subject_id: number
  chapters: ChapterHeatmapItem[]
}

export interface QuestionTypeItem {
  question_type: string
  count: number
  percentage: number
}

export interface QuestionTypeDistribution {
  subject_id: number
  total: number
  items: QuestionTypeItem[]
}

export interface ScoreTrendValue {
  session_id: string
  exam_id: number
  exam_name: string
  date?: string
  subject_id: number
  subject_name: string
  score: number
  total_score: number
  score_rate: number
  mode: string
}

export interface SubjectScoreTrend {
  subject_id: number
  subject_name: string
  sessions: number
  average_score_rate: number
  best_score_rate: number
  latest_score_rate: number
  trend_slope: number
  trend: string
  values: Array<{ date?: string; score_rate: number }>
}

export interface ChapterScoreTrend {
  chapter_id: number
  chapter_name: string
  total_count: number
  wrong_count: number
  correct_count: number
  score_rate: number
}

export interface ScoreTrendSummary {
  sessions: number
  average_score_rate: number
  latest_score_rate: number
  best_score_rate: number
  trend_slope: number
  trend: string
}

export interface ScoreTrend {
  user_id: number
  subject_id?: number
  summary: ScoreTrendSummary
  values: ScoreTrendValue[]
  subjects: SubjectScoreTrend[]
  chapters: ChapterScoreTrend[]
}

export const analysisApi = {
  getKnowledgeTree: (subjectId: number) =>
    unwrap<KnowledgeTree>(apiClient.get(`/analysis/knowledge-tree/${subjectId}`)),
  getHighFrequencyPoints: (subjectId: number, limit = 20) =>
    unwrap<HighFrequencyPoint[]>(
      apiClient.get(`/analysis/high-frequency/${subjectId}`, { params: { limit } }),
    ),
  getPointTrend: (subjectId: number, pointId: number, years = 5) =>
    unwrap<PointTrend>(
      apiClient.get(`/analysis/trend/${subjectId}/${pointId}`, { params: { years } }),
    ),
  getWordCloud: (subjectId: number) =>
    unwrap<WordCloudItem[]>(apiClient.get(`/analysis/word-cloud/${subjectId}`)),
  getKnowledgeNetwork: (subjectId: number, limit = 30) =>
    unwrap<KnowledgeNetwork>(
      apiClient.get(`/analysis/knowledge-network/${subjectId}`, { params: { limit } }),
    ),
  getChapterHeatmap: (subjectId: number) =>
    unwrap<ChapterHeatmap>(apiClient.get(`/analysis/chapter-heatmap/${subjectId}`)),
  getQuestionTypeDistribution: (subjectId: number) =>
    unwrap<QuestionTypeDistribution>(apiClient.get(`/analysis/question-types/${subjectId}`)),
  predictNextExam: (subjectId: number) =>
    unwrap<HighFrequencyPoint[]>(apiClient.get(`/analysis/prediction/${subjectId}`)),
  getHotspots: (subjectId: number) =>
    unwrap<HotspotAlert[]>(apiClient.get(`/analysis/hotspots/${subjectId}`)),
  getScoreTrends: (userId: number, subjectId?: number, limit = 20) =>
    unwrap<ScoreTrend>(
      apiClient.get(`/analysis/score-trends/${userId}`, {
        params: { subject_id: subjectId, limit },
      }),
    ),
  aiAnalyzeHotspots: (subjectId: number) =>
    unwrap<AIHotspotAnalysis>(apiClient.get(`/analysis/ai-hotspots/${subjectId}`, { timeout: 60000 })),
  getAnswerTemplates: () =>
    unwrap<AnswerTemplatesResult>(apiClient.get('/analysis/templates')),
  getTemplateByType: (questionType: string, category = 'default') =>
    unwrap<AnswerTemplate>(
      apiClient.get(`/analysis/templates/${questionType}`, { params: { category } }),
    ),
  getSprintReport: (userId: number, subjectId: number, examDate?: string) =>
    unwrap<SprintReport>(
      apiClient.get(`/analysis/sprint-report/${userId}/${subjectId}`, {
        params: examDate ? { exam_date: examDate } : {},
      }),
    ),
}

export interface AIHotspotPrediction {
  name: string
  confidence: number
  reason: string
  study_tip: string
}

export interface AIHotspotAnalysis {
  subject_id: number
  subject_name: string
  ai_powered: boolean
  summary: string
  top_predictions: AIHotspotPrediction[]
  weak_areas: string[]
  study_plan: string[]
  exam_tips: string[]
  raw_hotspots: HotspotAlert[]
  raw_high_points: HighFrequencyPoint[]
}

// Answer template types
export interface AnswerTemplate {
  name: string
  question_type: string
  question_type_label: string
  category: string
  structure: string
  scoring_tips: string[]
  example_frame: string
  common_mistakes: string[]
  score_range?: string
}

export interface AnswerTemplateGroup {
  question_type: string
  label: string
  templates: AnswerTemplate[]
}

export interface AnswerTemplatesResult {
  fill_blank?: AnswerTemplateGroup
  short_answer?: AnswerTemplateGroup
  essay?: AnswerTemplateGroup
  case?: AnswerTemplateGroup
  exam_tips: {
    time_management: string
    answer_order: string
    key_reminders: string[]
  }
}

// Sprint report types
export interface SprintReport {
  generated_at: string
  subject_id: number
  exam_date: string | null
  days_remaining: number | null
  current_assessment: {
    estimated_score: number
    pass_probability: {
      level: string
      label: string
      percent: number
      description: string
    }
    recent_trend: string
    total_practice_count: number
    unmastered_wrong_count: number
  }
  pass_strategy: {
    target_score: number
    priority_order: string[]
    focus_advice: string[]
    guaranteed_minimum: string
  }
  review_checklist: {
    must_review_points: Array<{
      point_id: number
      name: string
      frequency: number
      importance: string
      mastery_level: number
      priority: string
    }>
    type_weakness: Array<{
      question_type: string
      label: string
      total_count: number
      correct_count: number
      accuracy: number
      score_rate: number
    }>
    unmastered_wrong_count: number
  }
  time_allocation: {
    strategy: string
    details: string
    allocation: Array<{ item: string; percent: number }>
  }
  exam_tips: {
    time_management: string
    answer_order: string
    key_reminders: string[]
  }
}

File: backend/api/analysis.py
# -*- coding: utf-8 -*-
"""Analysis API routes."""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from backend.database import get_db
from backend.api.dependencies import get_current_user, require_self
from backend.models.user import User
from backend.services.analysis_service import AnalysisService
from backend.config import settings
from backend.utils.rate_limit import rate_limiter
from backend.services.ai_provider_pool import ai_provider_pool

router = APIRouter(prefix="/analysis", tags=["analysis"])


@router.get("/ai-providers/status")
async def get_ai_provider_status(current_user: User = Depends(get_current_user)):
    """Return non-secret provider pool and health information."""
    return {"code": 0, "data": ai_provider_pool.public_status()}


@router.get("/knowledge-tree/{subject_id}")
async def get_knowledge_tree(subject_id: int, db: Session = Depends(get_db)):
    return {"code": 0, "data": AnalysisService(db).get_knowledge_tree(subject_id)}


@router.get("/high-frequency/{subject_id}")
async def get_high_frequency_points(
    subject_id: int,
    limit: int = Query(default=20, ge=1, le=100),
    db: Session = Depends(get_db),
):
    return {"code": 0, "data": AnalysisService(db).get_high_frequency_points(subject_id, limit)}


@router.get("/trend/{subject_id}/{point_id}")
async def get_point_trend(
    subject_id: int,
    point_id: int,
    years: int = Query(default=5, ge=1, le=20),
    db: Session = Depends(get_db),
):
    return {"code": 0, "data": AnalysisService(db).get_point_trend(subject_id, point_id, years)}


@router.get("/word-cloud/{subject_id}")
async def get_word_cloud(subject_id: int, db: Session = Depends(get_db)):
    return {"code": 0, "data": AnalysisService(db).get_word_cloud_data(subject_id)}


@router.get("/knowledge-network/{subject_id}")
async def get_knowledge_network(
    subject_id: int,
    limit: int = Query(default=30, ge=1, le=100),
    db: Session = Depends(get_db),
):
    return {"code": 0, "data": AnalysisService(db).get_knowledge_network(subject_id, limit)}


@router.get("/chapter-heatmap/{subject_id}")
async def get_chapter_heatmap(subject_id: int, db: Session = Depends(get_db)):
    return {"code": 0, "data": AnalysisService(db).get_chapter_heatmap(subject_id)}


@router.get("/question-types/{subject_id}")
async def get_question_type_distribution(subject_id: int, db: Session = Depends(get_db)):
    return {"code": 0, "data": AnalysisService(db).get_question_type_distribution(subject_id)}


@router.get("/prediction/{subject_id}")
async def predict_next_exam(subject_id: int, db: Session = Depends(get_db)):
    return {"code": 0, "data": AnalysisService(db).predict_next_exam(subject_id)}


@router.get("/hotspots/{subject_id}")
async def get_hotspot_alerts(subject_id: int, db: Session = Depends(get_db)):
    return {"code": 0, "data": AnalysisService(db).get_hotspot_alerts(subject_id)}


@router.get("/score-trends/{user_id}")
async def get_score_trends(
    user_id: int,
    subject_id: int | None = Query(default=None, ge=1),
    limit: int = Query(default=20, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    require_self(user_id, current_user)
    return {"code": 0, "data": AnalysisService(db).get_score_trends(user_id, subject_id, limit)}


@router.get("/ai-hotspots/{subject_id}")
def ai_analyze_hotspots(
    subject_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """AI 智能分析高频考点，生成备考建议和考点预测。

    从本地配置文件读取 AI API 配置，未配置时回退到纯统计分析。
    """
    rate_limiter.check("ai_analysis", f"user:{current_user.id}", settings.AI_RATE_LIMIT_PER_MINUTE)
    return {"code": 0, "data": AnalysisService(db).ai_analyze_hotspots(subject_id)}


@router.get("/templates")
async def get_answer_templates():
    """获取所有答题模板"""
    from backend.services.answer_template_service import AnswerTemplateService
    service = AnswerTemplateService()
    return {"code": 0, "data": service.get_all_templates()}


@router.get("/templates/{question_type}")
async def get_template_by_type(
    question_type: str,
    category: str = Query(default="default"),
):
    """获取某题型的答题模板"""
    from backend.services.answer_template_service import AnswerTemplateService
    service = AnswerTemplateService()
    template = service.get_template(question_type, category)
    if not template:
        from fastapi import HTTPException
        raise HTTPException(status_code=404, detail="模板不存在")
    return {"code": 0, "data": template}


@router.get("/sprint-report/{user_id}/{subject_id}")
async def get_sprint_report(
    user_id: int,
    subject_id: int,
    exam_date: str | None = Query(default=None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """生成考前冲刺报告"""
    require_self(user_id, current_user)
    from datetime import datetime
    from backend.services.sprint_report_service import SprintReportService
    service = SprintReportService(db)
    exam_dt = datetime.fromisoformat(exam_date) if exam_date else None
    return {"code": 0, "data": service.generate_report(user_id, subject_id, exam_dt)}
